### PR TITLE
[WIP] Add clang-cl build support for Windows

### PR DESCRIPTION
**Pull request by peterjc123**
In order to enable OpenMP 2.0+ functions on Windows, this PR is made. However, we need to change some flags in ONNX and add build tools on CI for this PR to pass.

### CODE REVIEW

**Comment by peterjc123** [comment]
@yf225 Could you please deploy Clang on the Windows CI machine?
@ezyang How could I change onnx code? Just submit here or to the onnx repo?

**Comment by peterjc123** [comment]
When using Clang, we have to append the following flags to `onnx/CMakeLists.txt` after the use of  `/WX`:
```
-Wno-implicit-function-declaration
-Wno-undefined-inline
-Wno-incompatible-pointer-type
-Wno-dllexport-explicit-instantiation-decl
-Wno-microsoft-unqualified-friend
-Wno-absolute-value
-Wno-writable-strings
-Qunused-arguments
```.
The flag `/MP` should also be replaced by `-Xclang -fopenmp`.  However, by doing this, the intel-openmp libraries like `libomp5md.lib` are now dependencies.

**Comment by peterjc123** [comment]
The full log of the building process is available [here](https://gist.github.com/peterjc123/85d9fb8a4036407c6470f7677b8d4787). I also ran the tests on my dev machine and it passed.

**Comment by ezyang** [comment]
@peterjc123 You'll have to open a PR in the ONNX repo.

**Comment by peterjc123** [comment]
It looks like clang-cl cannot play happily with pybind11 and is causing random crashes. So is there a method to specify generators for certain source files? We just compile the code under TH and THNN using clang-cl and keep the others going through MSVC. In that way, we don't need to modify ONNX and any other libraries that don't play well with clang-cl. What do you think? @yf225

**Comment by yf225** [comment]
Do we know why the random crashes happened? It would be nice to keep the toolchain simple, but maybe compiling them separately is the only feasible way. @ezyang Thoughts on this?

**Comment by ezyang** [comment]
Yeah, I'd prefer it if we spent some time diagnosing why the crashes happen. Can you post some examples where this happened?

**Comment by peterjc123** [comment]
@ezyang Some failed test items can be seen [here](https://github.com/onnx/onnx/pull/1138#issuecomment-399731573) and [there](https://ci.appveyor.com/project/onnx/onnx/build/0.3.4055).

**Comment by ezyang** [comment]
Wow, these logs are supremely unhelpful :> I guess we'll have to run them ourselves...

**Comment by peterjc123** [comment]
@ezyang The issue in pybind11 may be helpful: https://github.com/pybind/pybind11/issues/1440.

**Comment by peterjc123** [comment]
@yf225 @ezyang Since the upstream PR is merged, should we try to test it here?

**Comment by yf225** [comment]
@peterjc123 I am having some issue with deploying a new Windows image for AWS (the new machines don't connect to Jenkins). Is there a way to download the `clang` toolchain in the build script in order to test this PR? I have the following to download and install LLVM:
```
curl.exe -k -O https://s3.amazonaws.com/ossci-windows/LLVM-6.0.0.7z
7z x ./LLVM-6.0.0.7z -o"C:\Program Files\LLVM"
cmd.exe /c "C:\Program Files\LLVM\tools\msbuild\install.bat"
```
In the mean time I will figure out how to make the new images work with Jenkins.

**Comment by peterjc123** [comment]
Now the problem is on the compatibitility issues of clang-cl and sccache caused by
cmake.

File: .jenkins/pytorch/win-build.sh
Comment (peterjc123) [inline .jenkins/pytorch/win-build.sh]: I'm not sure whether we should change all windows builds to use clang-cl or set up a new build spec.

**Comment by yf225** [comment]
@peterjc123 clang-cl with sccache now works locally and I've uploaded the new sccache binary. We will need to change the LLVM download file to `LLVM-6.0.1-rC331533.7z`, and then change the `set CC=sccache cl` and `set CXX=sccache cl` lines in the original build script to `set CC=sccache clang-cl` and `set CXX=sccache clang-cl`.

**Comment by peterjc123** [comment]
@yf225 Great! I'll try it out.

File: .jenkins/pytorch/win-build.sh
Comment (yf225) [inline .jenkins/pytorch/win-build.sh]: Does `clang-cl` only work for CPU builds and not CUDA builds? If not we should probably change the `set CC=sccache cl` and `set CXX=sccache cl` lines above instead.

File: .jenkins/pytorch/win-build.sh
Comment (yf225) [inline .jenkins/pytorch/win-build.sh]: Why do we remove this line?

File: .jenkins/pytorch/win-build.sh
Comment (yf225) [inline .jenkins/pytorch/win-build.sh]: We should change this to `LLVM-6.0.1-rC331533.7z` as well.

File: .jenkins/pytorch/win-build.sh
Comment (yf225) [inline .jenkins/pytorch/win-build.sh]: It would be great to add `-aoa` options to 7z here (similar to how we do it for magma above), so that existing files in `C:\Program Files\LLVM` are guaranteed to be rewritten. Otherwise local builds with this script might run into a prompt that asks whether to overwrite the existing files.

**Review by yf225 — CHANGES_REQUESTED** [review]
See comments

File: .jenkins/pytorch/win-build.sh
Comment (peterjc123) [reply .jenkins/pytorch/win-build.sh]: @yf225 It will fallback to cl.exe.

**Comment by SsnL** [comment]
@yf225 your comments seem to have been resolved :). Could you review when you have time?

**Comment by yf225** [comment]
@peterjc123 Do you know how we can fix this error:
```
23:27:46 [421/497] Linking CXX shared library bin\caffe2.dll
23:27:46 FAILED: bin/caffe2.dll LIB/caffe2.lib 
23:27:46 cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=caffe2\CMakeFiles\caffe2.dir --manifests  -- C:\PROGRA~2\MIB055~1\2017\COMMUN~1\VC\Tools\MSVC\1411~1.255\bin\Hostx86\x64\link.exe /nologo @CMakeFiles/caffe2.rsp  /out:bin\caffe2.dll /implib:LIB\caffe2.lib /pdb:bin\caffe2.pdb /dll /version:0.0 /INCREMENTAL:NO /NODEFAULTLIB:vcomp   && cd ."
23:27:46 LINK: command "C:\PROGRA~2\MIB055~1\2017\COMMUN~1\VC\Tools\MSVC\1411~1.255\bin\Hostx86\x64\link.exe /nologo @CMakeFiles/caffe2.rsp /out:bin\caffe2.dll /implib:LIB\caffe2.lib /pdb:bin\caffe2.pdb /dll /version:0.0 /INCREMENTAL:NO /NODEFAULTLIB:vcomp /MANIFEST /MANIFESTFILE:bin\caffe2.dll.manifest" failed (exit code 1120) with the following output:
23:27:46    Creating library LIB\caffe2.lib and object LIB\caffe2.exp
23:27:46 OptionsGuard.cpp.obj : error LNK2019: unresolved external symbol "public: __int64 __cdecl at::Tensor::get_device(void)const " (?get_device@Tensor@at@@QEBA_JXZ) referenced in function "public: __cdecl at::TensorOptions::TensorOptions(struct at::Tensor,bool)" (??0TensorOptions@at@@QEAA@UTensor@1@_N@Z)
23:27:46 TensorOptions.cpp.obj : error LNK2001: unresolved external symbol "public: __int64 __cdecl at::Tensor::get_device(void)const " (?get_device@Tensor@at@@QEBA_JXZ)
23:27:46 bin\caffe2.dll : fatal error LNK1120: 1 unresolved externals
23:27:46 ninja: build stopped: subcommand failed.
```

**Comment by peterjc123** [comment]
@yf225 May I know where is this one defined?

**Comment by yf225** [comment]
@peterjc123 Do you mind doing it in this PR? I don't have push access to this PR, but it would be better than opening another PR because it would be more obvious what it fixes. Thanks!

**Comment by yf225** [comment]
@peterjc123 I think that the `clang-cl.exe: error: no such file or directory: 'NOTFOUND'` error is because of this: https://github.com/dmlc/xgboost/issues/2596#issuecomment-323175810, although I am not sure whether it's caused by OpenMP or not

**Comment by peterjc123** [comment]
@yf225 I found out the reason in the log. It says:

> -- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Found OpenMP_CXX: -Xclang -fopenmp  

So it'll be fine after we set OpenMP_C_FLAGS.

**Comment by peterjc123** [comment]
The test `test_default_arguments_are_used` in `test_jit.py` seem to be failing in clang-cl builds.

**Comment by peterjc123** [comment]
Currently, it was blocked by the missing support for `__int128` division in clang-cl, which will be added in this [PR](https://reviews.llvm.org/D41813).

**Comment by peterjc123** [comment]
Continued in https://github.com/pytorch/pytorch/pull/63288.

**Pull request closed without merging**
